runtime/runtime/src/ethereum.rs

```diff
@@ -0,0 +1,161 @@
+        let mut path = cache_dir.clone();
+        path.push(format!("{}", epoch));
+        path
+    }
+
+    pub fn new(cache_dir: &PathBuf, epoch: u64) -> Self {
+        match Self::from_file(cache_dir, epoch) {
+            Ok(light_cache) => light_cache,
+            Err(_) => {
+                let cache_size = ethash::get_cache_size(epoch as usize);
+                let mut cache = Vec::with_capacity(cache_size);
+                cache.resize(cache_size, 0);
+                let seed = ethash::get_seedhash(epoch as usize);
+                ethash::make_cache(&mut cache, seed);
+                let _ = Self::to_file(cache_dir, epoch, &cache);
+                LightCache { epoch, cache }
+            }
+        }
+    }
+
+    fn to_file(cache_dir: &PathBuf, epoch: u64, cache: &[u8]) -> io::Result<usize> {
+        create_dir_all(cache_dir)?;
+        let mut file = File::create(Self::storage_file(cache_dir, epoch))?;
+        file.write(cache)
+    }
+
+    pub fn from_file(cache_dir: &PathBuf, epoch: u64) -> io::Result<Self> {
+        let mut file = File::open(Self::storage_file(cache_dir, epoch))?;
+        let mut cache: Vec<u8> = Vec::with_capacity(
+            file.metadata().map(|m| m.len() as _).unwrap_or(NODE_BYTES * 1_000_000),
+        );
+        file.read_to_end(&mut cache)?;
+        cache.shrink_to_fit();
+
+        if cache.len() % NODE_BYTES != 0 || cache.capacity() % NODE_BYTES != 0 {
+            return Err(io::Error::new(
+                io::ErrorKind::Other,
+                "Node cache is not a multiple of node size",
+            ));
+        }
+        Ok(LightCache { epoch, cache })
+    }
+}
+
+pub struct EthashProvider {
+    cache_dir: PathBuf,
+    recent: Option<LightCache>,
+    prev: Option<LightCache>,
+}
+
+impl EthashProvider {
+    pub fn new(cache_dir: &Path) -> Self {
+        EthashProvider { cache_dir: cache_dir.to_path_buf(), recent: None, prev: None }
+    }
+
+    pub fn check_ethash(
+        &mut self,
+        block_number: u64,
+        header_hash: H256,
+        nonce: H64,
+        mix_hash: H256,
+        difficulty: U256,
+    ) -> bool {
+        let epoch = block_number / ETHASH_EPOCH_LENGTH;
+        let cache = {
+            let cache = match &self.recent {
+                Some(recent) if recent.epoch == epoch => self.recent.as_ref(),
+                _ => match self.prev {
+                    Some(ref prev) if prev.epoch == epoch => {
+                        if self.recent.is_some() && self.recent.as_ref().unwrap().epoch > prev.epoch
+                        {
+                            None
+                        } else {
+                            mem::swap(&mut self.prev, &mut self.recent);
+                            self.recent.as_ref()
+                        }
+                    }
+                    _ => None,
+                },
+            };
+            match cache {
+                None => {
```

Comment:
This implementation can easily be abused by using random block_number. Overall we shouldn't add it without security considerations. 
Now if your node needs to catch up you can't rely on only having recent block_numbers, but it may be requesting really old blocks.
If computing cache takes 25 seconds, then your node might just go completely down after a single block.

Reply:
I am confused. The block number here is ethereum block number, not our block number. For a random block number, we will reject it if it is not captured by the cache (i.e, too far in the future). Also the cache only needs to be computed every 30000 blocks.

Reply:
There are some features that have not been implemented yet (for example tracking ethereum headers), which might be the point of confusion.

Reply:
Yes. I don't see we rejecting random headers yet. So how about we don't expose it until we do